source/grammar/index.rst

```diff
@@ -80,53 +88,59 @@ repetition is denoted by curly brackets ``+`` (one or more) and ``*`` (zero or m
         :| ">" | "<" | ">=" | "<=" | "==" | "!="
     assignment_operator: "=" | "+=" | "-=" | "*=" | "/=" |
         :| "<<=" | ">>=" |  "&=" | "|=" | "^=" | "~=" | "->"
     membership_test: id "in" set_declaration
-    index_set_declaration: "{" index_set_generators "}"
-    index_set_generators: expression_list | range
-    range: expression? ":" expression? ( ":" expression )?
+    index_set_declaration: "{" expression_list "}" | range
+    range: "[" expression? ":" expression? ( ":" expression )? "]"
     selection_statement: "if (" expression ")" program_block ( "else" program_block )?
-    iteration_statement: "for" membership_test program_block | "while (" expression ")" program_block
+    iteration_statement: ( "for" membership_test | "while (" expression ")" ) program_block
     selection_directive_statement: selection_directive ";"
     selection_directive: "break" | "continue" | "exit"
     quantum_gate_definition: "gate" quantum_gate_signature quantum_gate_block
-    quantum_gate_signature: id quantum_gate_args id_lis
-    quantum_gate_args: ( "(" classical_type_and_id_list? ")" )?
+    quantum_gate_signature: id classical_params id_list
+    id_list: ( id "," )* id
+    index_type: id ( "[" expression "]" )?
+    index_list: ( index_type "," ) index_type
+    classical_params: ( "(" classical_type_and_id_list? ")" )?
     quantum_gate_block: "{" quantum_gate_call* "}"
     quantum_type_and_id_list: ( quantum_type_and_id "," )* quantum_type_and_id
-    quantum_type_and_id: quantum_type designator? association 
+    quantum_type_and_id: quantum_type designator? association
     quantum_statement: quantum_instruction ";"
     quantum_instruction: quantum_gate_call
         :| quantum_measurement
-    quantum_measurement: "measure" argument
+        :| barrier_instruction
+    quantum_measurement: "measure" quantum_declaration "->" bit_declaration
+        :| bit_declaration "= measure" quantum_declaration
+    barrier_instruction: "barrier" index_list
     quantum_gate_modifiers: ( "inv" | "pow" "[" nninteger "]" | "ctrl" ) "@"
-    quantum_gate_call: quantum_gate_name quantum_gate_args designator any_list ";"
-    quantum_gate_name: "CX" | "U" | "delay" | "reset" | id
+    quantum_gate_call: quantum_gate_name constant_args index_list | delay_call ";"
+    delay_call: "delay" ( "[" id | time_unit "]" )? ( range | index_list )
+    quantum_gate_name: "CX" | "U" | "reset" | id
         :| quantum_gate_modifier "@" quantum_gate_name
     quantum_gate_modifier: "inv" | "ctrl"
         :| "pow" nninteger?
-    quantum_gate_args: ( "(" expression_list? ")" )?
     quantum_declaration: quantum_type id designator
     quantum_type: "qubit" | "qreg"
     qubit_id_list: ( qubit_id "," )* qubit_id
-    qubit_id: id | "%" [ "q" ] nninteger
+    qubit_id: id | physical_qubit_id
+    physical_qubit_list: ( physical_qubit_id "," )* physical_qubit_id
+    physical_qubit_id: "%" [ "q" ] nninteger
```

Comment:
Why does this need to be this way? I don't think it does. Getting too specific with the grammar.

Reply:
It was my understanding that `defcal` requires using physical qubits. That's why this was added. The spec says
`The entry point to such gate and measurement definitions is the defcal keyword analogous to the gate keyword, but where the defcal body specifies a pulse-level instruction sequence on physical qubits`

Reply:
Within "Pulse-level descriptions of gates and measurement" section.

Reply:
It means logically the index is on assigned physical qubits, its still possible to define a qubit argument that a-priori we don't know which physical qubit it is. This is useful for example for gates that are always the same like Rz(theta) gates.

Reply:
Fixed in https://github.com/Qiskit/openqasm/pull/89